### PR TITLE
Jazz MediaRanker

### PR DESCRIPTION
**Pull request by jazziesf**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote two methods one to sort all media by votes for each category and another to sort all media for the spotlight.
Describe how you approached testing that model method. What edge cases did you come up with? | If work was nil, user was not logged in to vote or create work, vote more than once. 
What are `session` and `flash`? What is the difference between them? | sessions is a way to keep track of users "session when they are in a program. Flash sends a one-time messages from controllers to your views.
Describe a controller filter you wrote. | find_by user control filter since each method had to look for a user by it's id. 
What was one thing that you gained more clarity on through this assignment? | flash and flash now, what the model knows and how it sends information to the controller. 
What is the Heroku URL of your deployed application? | https://mediarankerada-app.herokuapp.com/ 
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  resources :users
+  post 'users/login', to: 'users#login', as: 'login'
+
+  # resources :works
+  resources :works do
+    resources :votes, only: [:create]
+  end
+
+  post 'works/:id/upvote', to: 'works#upvote', as: 'upvote'
```
Comment (CheezItMan) [inline config/routes.rb:15]: If you have `works#upvote`, why do you have `votes#create`.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class VotesController < ApplicationController
+
+  def create
+    id = params[:work_id].to_i
+    @work = Work.find_by(id: id)
+
+    if @current_user
+    @work.upvote(user_id: @current_user.id)
```
Comment (CheezItMan) [inline app/controllers/votes_controller.rb:8]: indentation

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    @albums = Work.list_by_votes("album")
+    @movies = Work.list_by_votes("movie")
+
+  end
+
+  def show
+    id = params[:id]
+    @work = Work.find_by(id: id)
+
+    if id == nil
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:14]: This should probably be:

```ruby
if @work.nil?
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+  def index
+    @users = User.all.order(:id)
+  end
+
+  def show
+    id = params[:id]
+    @user = User.find_by(id: id)
+
+    if id == nil
```
Comment (CheezItMan) [inline app/controllers/users_controller.rb:12]: this should be

```ruby
if @user.nil?
```

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <link rel="shortcut icon" type="image/x-icon" href="/assets/owl.jpg">
+
+  </head>
+
+  <body>
+    <header class="app-header__header">
+        <h1>
+          <a href="/">Media Ranker</a>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:17]: You could also use `link_to` here.

File: app/views/works/main.html.erb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  </section>
+
+<hr class="root__hr">
+
+<section class="top-ten__container">
+  <section class="top-ten__list-container">
+  <h2 class="top-ten__header">Top Movies</h2>
+  <ul class="list-group top-ten__list">
+    <% @movies.first(10).each do |work| %>
```
Comment (CheezItMan) [inline app/views/works/main.html.erb:19]: Notice how these tables are similar.  You could DRY up your code with partials to generate these tables.

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      vote = Vote.new(work_id: self.id, user_id: arg[:user_id])
+      self.votes << vote
+      vote.save
+    else
+      raise ArugmentError, "User Id doesnt exsist"
+    end
+  end
+
+  def self.list_by_votes(category)
+    Work.all.sort_by {|work| -work.votes.count}.select { |work| work.category == category }
```
Comment (CheezItMan) [inline app/models/work.rb:22]: Nice use of the negative sign to sort them!

File: app/controllers/works_controller.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    elsif @current_user.nil?
+      flash[:warning] = "You must be logged in to vote"
+      redirect_to works_path
+    end
+  end
+
+
+  def destroy
+    id = params[:id]
+    @work = Work.find_by(id: id)
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:107]: You should go through all the votes for this work and destroy them first.  
```ruby
@work.votes.each do |vote|
  vote.destory
end
```

Right now I'm unable to delete works.

File: test/models/user_test.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      user = users(:jazz)
+      expect(user.valid?).must_equal true
+    end
+
+    it 'is an invalid username name blank' do
+      user = users(:blank)
+      expect(user.valid?).must_equal false
+    end
+
+    it 'is an invalid length for a username' do
```
Comment (CheezItMan) [inline test/models/user_test.rb:22]: You should also test the edge-cases for username length, test with length 4 and test with length 5.

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    it 'is an invalid username duplicate' do
+      user = users(:jazz)
+      user = User.new(username: "jazztaz")
+      expect(user.valid?).must_equal false
+    end
+
+  end
+
+  describe 'relations' do
+      before(:each) do
```
Comment (CheezItMan) [inline test/models/user_test.rb:36]: This before isn't actually doing anything.

File: test/models/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        user = users(:jazz)
+        work = works(:book2)
+        user.works << work
+      end
+
+      it "should have created a relationship" do
+        Vote.first.user.must_be_instance_of User
```
Comment (CheezItMan) [inline test/models/user_test.rb:43]: This is more testing `Vote` than `User`

File: test/models/vote_test.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  let (:jazz) {users(:jazz)}
+  let (:vote1) {votes(:vote1)}
+  let (:book) {works(:book)}
+  let (:album) {works(:album)}
+
+  it 'creates a valid vote' do
+    expect(vote1.valid?).must_equal true
+  end
+
+  it "adds a new vote to work" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:13]: This test is really testing if ActiveRecord functionality.

File: test/models/vote_test.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+      expect(new_vote.errors.messages).must_equal :user=>["must exist"]
+    end
+
+    it "is invalid without a work id" do
+      new_vote = Vote.create(user:jazz)
+      expect(new_vote.valid?).must_equal false
+      expect(new_vote.errors.messages).must_equal :work=>["must exist"]
+    end
+
+    it "is invalid with more than one vote" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:35]: Invalid with more than one vote for the same work by the same person.

File: test/models/work_test.rb [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+    it 'is invalid if title is not unique' do
+      work = works(:book)
+      work = Work.new(title: "Harry Potter and the Goblet of Fire" )
+
+      expect(work.valid?).must_equal false
+      expect(work.errors.messages).must_include :title
+    end
+
+    it 'is invalid if title is not longer than 5 char' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:31]: You should test it with 4 characters and with 5, to verify the edge-cases.

File: test/models/work_test.rb [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+      expect(work.errors.messages).must_include :publication_year
+    end
+
+    it 'is invalid if description is blank' do
+      work = works(:publication)
+      expect(work.valid?).must_equal false
+      expect(work.errors.messages).must_include :description
+    end
+
+    it 'is invalid if publication_year is less than 999 and greater than 10000' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:55]: 💯

File: test/models/work_test.rb [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+      let (:book3) {works(:book3)}
+      let (:album1) {works(:album1)}
+      let (:blank) {works(:blank)}
+
+      it "sorts work by vote count" do
+        expect(Work.list_by_votes("book").first.title).must_equal book2.title
+        expect(Work.list_by_votes("book").last.title).must_equal book.title
+        expect(Work.list_by_votes("album").first.title).must_equal album1.title
+      end
+
```
Comment (CheezItMan) [inline test/models/work_test.rb:90]: No test for `spotlight`, and no test for `upvote`.

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages
Comprehension questions	|  check, glad you gained some clarity
**General** |
Rails fundamentals (RESTful routing, use of named paths) |   Check, You do have 2 different ways to create votes, that was odd.  
Views are well-organized (DRY, use of semantic HTML, use of partials) |  You could DRY up your views with partials, but otherwise well done.  
Errors are reported to the user |  Check
Business logic lives in the models | Check
Models are thoroughly tested, including relations, validations and any custom logic |  Check
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |  Check
**Wave 2 - Users and Votes** |
Users can log in and log out |  Check
The ID of the current user is stored in the session |  Check
A user cannot vote for the same media more than once |  Check
All media lists are ordered by vote count |  Check
Splash page contains a media spotlight |  Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users |  Check
Individual user pages and the user list are present |  Check
**Optional - Styling** |
Bootstrap is used appropriately |  Well done
Look and feel is similar to the original |  Nice work, the site looks great
**Overall** |  Great work, you hit all the learning goals.  You should look to improve on your tests.  Check out my inline notes and let me know if you have any questions.
